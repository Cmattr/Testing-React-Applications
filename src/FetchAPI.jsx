import React, { useState, useEffect } from "react";
import axios from "axios";

const FetchAPI = () => {
    const [posts, setPosts] = useState([]);
    const [newPost, setNewPost] = useState({ title: '', body: '' });
    const [editingPost, setEditingPost] = useState(null);

    // Function to fetch posts
    const fetchPosts = async () => {
        try {
            const response = await axios.get('https://jsonplaceholder.typicode.com/posts');
            setPosts(response.data);
        } catch (error) {
            console.error('Error fetching posts:', error);
        }
    };

    // Function to create a new post
    const createPost = async (e) => {
        e.preventDefault();
        try {
            const response = await axios.post('https://jsonplaceholder.typicode.com/posts', newPost);
            setPosts([...posts, response.data]); // Update state with the new post
            setNewPost({ title: '', body: '' }); // Clear the form
        } catch (error) {
            console.error('Error creating post:', error);
        }
    };

    // Function to handle edit
    const startEdit = (post) => {
        setEditingPost(post);
        setNewPost({ title: post.title, body: post.body });
    };

    // Function to update a post
    const updatePost = async (e) => {
        e.preventDefault();
        if (!editingPost) return;

        try {
            const response = await axios.put(`https://jsonplaceholder.typicode.com/posts/${editingPost.id}`, newPost);
            setPosts(posts.map(post => post.id === editingPost.id ? response.data : post)); // Update the specific post
            setEditingPost(null); // Clear the editing state
            setNewPost({ title: '', body: '' }); // Clear the form
        } catch (error) {
            console.error('Error updating post:', error);
        }
    };

    // Function to delete a post
    const deletePost = async (id) => {
        try {
            await axios.delete(`https://jsonplaceholder.typicode.com/posts/${id}`);
            setPosts(posts.filter(post => post.id !== id)); // Remove the deleted post from the state
        } catch (error) {
            console.error('Error deleting post:', error);
        }
    };

    // Fetch posts when component mounts
    useEffect(() => {
        fetchPosts();
    }, []);

    return (
        <div>
            <h1>Posts</h1>
            <ul>
                {posts.map(post => (
                    <li key={post.id}>
                        <h2>{post.title}</h2>
                        <p>{post.body}</p>
                        <button onClick={() => startEdit(post)}>Edit</button>
                        <button onClick={() => deletePost(post.id)}>Delete</button>
                    </li>
                ))}
            </ul>

            <h2>{editingPost ? "Edit Post" : "Create a New Post"}</h2>
            <form onSubmit={editingPost ? updatePost : createPost}>
                <div>
                    <label htmlFor="title">Title:</label>
                    <input
                        type="text"
                        id="title"
                        value={newPost.title}
                        onChange={(e) => setNewPost({ ...newPost, title: e.target.value })}
                        required
                    />
                </div>
                <div>
                    <label htmlFor="body">Body:</label>
                    <textarea
                        id="body"
                        value={newPost.body}
                        onChange={(e) => setNewPost({ ...newPost, body: e.target.value })}
                        required
                    ></textarea>
                </div>
                <button type="submit">{editingPost ? "Update Post" : "Add Post"}</button>
            </form>
        </div>
    );
};

export default FetchAPI;