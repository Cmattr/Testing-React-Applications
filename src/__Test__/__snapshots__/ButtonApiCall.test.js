import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import FetchAPI from "../../FetchAPI";
import axios from "axios";

jest.mock('axios');

describe("FetchAPI Component", () => {
    test('fetches and displays posts on component mount', async () => {
        const mockResponse = { data: [{ id: 1, title: "Test Post", body: "This is a test post." }] };
        axios.get.mockResolvedValue(mockResponse);

        render(<FetchAPI />);

        await waitFor(() => {
            expect(axios.get).toHaveBeenCalledWith('https://jsonplaceholder.typicode.com/posts');
            expect(screen.getByText("Test Post")).toBeInTheDocument();
            expect(screen.getByText("This is a test post.")).toBeInTheDocument();
        });
    });

    test('creates a new post and updates the UI', async () => {
        const mockPost = { id: 2, title: "New Post", body: "This is a new post." };
        axios.post.mockResolvedValue({ data: mockPost });

        const { getByLabelText, getByText, findByText } = render(<FetchAPI />);

        // Fill out the form
        fireEvent.change(getByLabelText(/Title:/i), { target: { value: "New Post" } });
        fireEvent.change(getByLabelText(/Body:/i), { target: { value: "This is a new post." } });
        fireEvent.click(getByText(/Add Post/i));

        await findByText("New Post");

        expect(axios.post).toHaveBeenCalledWith('https://jsonplaceholder.typicode.com/posts', { title: "New Post", body: "This is a new post." });
        expect(screen.getByText("New Post")).toBeInTheDocument();
        expect(screen.getByText("This is a new post.")).toBeInTheDocument();
    });

    test('edits a post and updates the UI', async () => {
        const initialPosts = [{ id: 1, title: "Old Post", body: "This is an old post." }];
        const updatedPost = { id: 1, title: "Updated Post", body: "This post has been updated." };

        axios.get.mockResolvedValue({ data: initialPosts });
        axios.put.mockResolvedValue({ data: updatedPost });

        render(<FetchAPI />);

        await waitFor(() => expect(screen.getByText("Old Post")).toBeInTheDocument());

        // Start editing
        fireEvent.click(screen.getByText(/Edit/i));
        fireEvent.change(screen.getByLabelText(/Title:/i), { target: { value: "Updated Post" } });
        fireEvent.change(screen.getByLabelText(/Body:/i), { target: { value: "This post has been updated." } });
        fireEvent.click(screen.getByText(/Update Post/i));

        await waitFor(() => {
            expect(axios.put).toHaveBeenCalledWith('https://jsonplaceholder.typicode.com/posts/1', { title: "Updated Post", body: "This post has been updated." });
            expect(screen.getByText("Updated Post")).toBeInTheDocument();
            expect(screen.getByText("This post has been updated.")).toBeInTheDocument();
        });
    });

    test('deletes a post and updates the UI', async () => {
        const initialPosts = [{ id: 1, title: "Post to Delete", body: "This post will be deleted." }];
        axios.get.mockResolvedValue({ data: initialPosts });
        axios.delete.mockResolvedValue({});

        render(<FetchAPI />);

        await waitFor(() => expect(screen.getByText("Post to Delete")).toBeInTheDocument());

        
        const deleteButtons = await screen.findAllByText(/Delete/i);
        expect(deleteButtons.length).toBe(1); 

        fireEvent.click(deleteButtons[0]); 

        await waitFor(() => {
            expect(axios.delete).toHaveBeenCalledWith('https://jsonplaceholder.typicode.com/posts/1');
            expect(screen.queryByText("Post to Delete")).not.toBeInTheDocument();
        });
    });

    test('matches the snapshot', () => {
        const { asFragment } = render(<FetchAPI />);
        expect(asFragment()).toMatchSnapshot();
    });
});
