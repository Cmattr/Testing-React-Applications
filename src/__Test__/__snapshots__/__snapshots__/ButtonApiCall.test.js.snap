// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchAPI Component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Posts
    </h1>
    <ul />
    <h2>
      Create a New Post
    </h2>
    <form>
      <div>
        <label
          for="title"
        >
          Title:
        </label>
        <input
          id="title"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="body"
        >
          Body:
        </label>
        <textarea
          id="body"
          required=""
        />
      </div>
      <button
        type="submit"
      >
        Add Post
      </button>
    </form>
  </div>
</DocumentFragment>
`;
